deeplake/core/chunk_engine.py

```diff
@@ -714,63 +722,149 @@ def _samples_to_chunks(
+            )
+            current_chunk._update_tensor_meta_length = False
             updated_chunks.append(current_chunk)
+            if extending:
+                enc_ids.append(current_chunk.id)
+        elif extending:
+            enc_ids.append(None)
         enc = self.chunk_id_encoder
         tiles = {}
-        nsamples = len(samples)
         if register and update_commit_diff:
             commit_diff = self.commit_diff
         if progressbar:
             pbar = tqdm(total=len(samples))
+        if not isinstance(samples, list) and not (
+            isinstance(samples, np.ndarray) and self._numpy_extend_optimization_enabled
+        ):
+            # Note: in the future we can get rid of this conversion of sample compressed chunks too by predicting the compression ratio.
+            samples = list(samples)
         while len(samples) > 0:
             num_samples_added = current_chunk.extend_if_has_space(
-                samples, update_tensor_meta=update_tensor_meta
+                samples, update_tensor_meta=update_tensor_meta, **extra_args
             )  # type: ignore
             if register_creds:
                 self.register_new_creds(num_samples_added, samples)
             if num_samples_added == 0:
-                current_chunk = self._create_new_chunk(register, row=start_chunk_row)
+                current_chunk = self._create_new_chunk(
+                    register and start_chunk_row is not None, row=start_chunk_row
+                )
+                current_chunk._update_tensor_meta_length = False
                 if start_chunk_row is not None:
                     start_chunk_row += 1
+                elif register:
+                    enc_ids.append(current_chunk.id)
+                    enc_count.append(0)
                 updated_chunks.append(current_chunk)
             elif num_samples_added == PARTIAL_NUM_SAMPLES:
                 sample = samples[0]
-                if register and sample.is_first_write:
-                    enc.register_samples(1)
-                if sample.is_last_write:
+                if sample.is_first_write:
                     if register:
-                        self.tile_encoder.register_sample(sample, self.num_samples - 1)
-                        if update_commit_diff:
-                            commit_diff.add_data(1)
-                    else:
-                        tiles[nsamples - len(samples)] = (
-                            sample.sample_shape,
-                            sample.tile_shape,
-                        )
+                        if start_chunk_row is not None:
+                            enc.register_samples(1)
+                        else:
+                            enc_count[-1] += 1
+                if sample.is_last_write:
+                    tiles[
+                        incoming_num_samples
+                        - len(samples)
+                        + bool(register) * orig_meta_length
```

Comment:
if register = True, we need absolute indices in tile encoder, with register = False, we need them relative.